src/main/scala/cromwell/binding/WdlExpression.scala

```diff
@@ -130,14 +129,23 @@ object WdlExpression {
             val message = failures.collect {case f: Failure[_] => f.exception.getMessage}.mkString("\n")
             Failure(new WdlExpressionException(s"Could not evaluate expression:\n$message"))
           case (successes, _) =>
-            successes.map{_.get.wdlType}.toSet match {
-              case s:Set[WdlType] if s.isEmpty =>
-                Failure(new WdlExpressionException(s"Can't have empty array declarations (can't infer type)"))
-              case s:Set[WdlType] if s.size == 1 =>
-                Success(WdlArray(WdlArrayType(s.head), successes.map{_.get}.toSeq))
-              case _ =>
-                Failure(new WdlExpressionException("Arrays must have homogeneous types"))
-            }
+            for (subtype <- WdlType.areTypesHomogeneous(successes.map(_.get)))
+              yield WdlArray(WdlArrayType(subtype), successes.map(_.get))
+        }
+      case a: Ast if a.getName == "MapLiteral" =>
+        val evaluatedMap = a.getAttribute("map").asInstanceOf[AstList].asScala.toVector map {kv =>
+          val key = evaluate(kv.asInstanceOf[Ast].getAttribute("key"), lookup, functions, interpolateStrings)
+          val value = evaluate(kv.asInstanceOf[Ast].getAttribute("value"), lookup, functions, interpolateStrings)
+          key -> value
+        }
+        val flattenedTries = evaluatedMap.flatMap {case (k,v) => Seq(k,v)}
+
+        flattenedTries.partition {_.isSuccess} match {
```

Comment:
`flattenedTries.partition(_.isSuccess)` of `flattenedTries partition { _is.Success }`